### DISCUSSION TITLE
Limit valid schema types to things that come from strings

### DISCUSSION THREAD

**Discussion by danni** [colinhacks/zod#2773] [Q&A] [answered]
I think this relates to #2421 and #2519 but I'm filing it separately because I'm not using `z.input` in any way.

I am looking to write a generic function for parsing data, which I know comes in the general form of:

```ts
{
  [K: string]: string | string[] | undefined
}
```

Some of those strings are of course other things, that I want to parse and transform. And so I'm trying to set up a generic function to validate that shape and do type coercion into a more useful shape. Where I want to limit the list of valid values to parsing functions that accept a string as input.

```ts
type ObjectSchema = Record<string, ObjectValue>
// Value must be optional or defaulted, let's flag our future parse error as a type error
type ObjectValue = z.ZodOptional<Inner1> | z.ZodDefault<Inner1>
// Value can either be a primitive or an array of primitives
type Inner1 = Inner2 | z.ZodArray<Inner2>
// Anything we can coerce from a string is fine
type Inner2 = z.ZodType<any, z.ZodTypeDef, string>

const objectSchema: ObjectSchema = {
  // arg0: z.string(), // Type Error, as expected
  arg1: z.string().optional(),
  arg2: z.string().default(""),
  arg3: z.coerce.number().default(0), // Type Error, is z.ZodNumber, knows nothing of its input type
  arg4: z.string().transform(val => parseInt(val, 10)).optional(), // Works, but not as pretty as the new syntax
  arg5: z.string().array().default([]),
}

type OutputType<T extends ObjectSchema> = z.infer<z.ZodObject<T>>

typeof OutputType<typeof objectSchema> = {
    arg1: string | undefined;
    arg2: string;
    arg3: number; // if it could be included
    arg4: number | undefined;
    arg5: string[];
}
```

I couldn't see anything obvious I could latch onto here to make this work using the new coerce options.

**Comment by JacobWeisenburger** [comment, marked as answer]
I don't fully understand what you are asking for. But perhaps this will help you:

```ts
const stringToNumber = z.string().pipe( z.coerce.number() )
type StringToNumberInput = z.input<typeof stringToNumber>
// type StringToNumberInput = string
type StringToNumberOutput = z.output<typeof stringToNumber>
// type StringToNumberOutput = number

const stringsToNumbers = stringToNumber.array()
type StringsToNumbersInput = z.input<typeof stringsToNumbers>
// type StringsToNumbersInput = string[]
type StringsToNumbersOutput = z.output<typeof stringsToNumbers>
// type StringsToNumbersOutput = number[]

```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by danni** [reply]
I had solutions using `pipe` and `transform`, but I was hoping for a way the type system could know from the coerce-func that it would be safe (the coerce-function works perfectly). It's just a convenience thing in this API I've been developing, where I know everything comes as a string, or array of strings, and so I wanted to lock down the list of valid Zod schemas to allow `z.coerce.number()` but not `z.number()` to catch the error sooner.

We could add a series of `zCoerceNumber`-type functions, but seeing as they would just be mimicing `z.coerce.number()` I was wondering if there was another way.

Thanks :)
